transformer_lens/head_detector.py

```diff
@@ -1,157 +1,237 @@
     detection_pattern = eq_mask.astype(int)
     return torch.tril(torch.as_tensor(detection_pattern).float())
 
+
 # Induction head
-def get_induction_head_detection_pattern(model: HookedTransformer, sequence: str) -> torch.Tensor:
-    """Outputs a detection score for induction heads. 
-    Induction head: https://dynalist.io/d/n2ZWtnoYHrU1s4vnFSAQ519J#z=_tFVuP5csv5ORIthmqwj0gSY
-    
+def get_induction_head_detection_pattern(
+    tokens: torch.Tensor,  # [batch (1) x pos]
+) -> torch.Tensor:
+    """Outputs a detection score for [induction heads](https://dynalist.io/d/n2ZWtnoYHrU1s4vnFSAQ519J#z=_tFVuP5csv5ORIthmqwj0gSY).
+
     Args:
-      model: Model being used.
-      sequence: String being fed to the model."""
-    
-    duplicate_pattern = get_duplicate_token_head_detection_pattern(model, sequence)
+      sequence: String being fed to the model.
+    """
+    duplicate_pattern = get_duplicate_token_head_detection_pattern(tokens)
 
     # Shift all items one to the right
     shifted_tensor = torch.roll(duplicate_pattern, shifts=1, dims=1)
 
     # Replace first column with 0's - we don't care about bos but shifting to the right moves the last column to the first, and the last column might contain non-zero values.
-    zeros_column = torch.zeros((duplicate_pattern.shape[0], 1))
+    zeros_column = torch.zeros(duplicate_pattern.shape[0], 1)
     result_tensor = torch.cat((zeros_column, shifted_tensor[:, 1:]), dim=1)
     return torch.tril(result_tensor)
 
-def get_supported_heads():
+
+def get_supported_heads() -> None:
     """Returns a list of supported heads."""
-    print(f"Supported heads: {HEAD_NAMES}")
\ No newline at end of file
+    print(f"Supported heads: {HEAD_NAMES}")
+
+
+def compute_head_attention_similarity_score(
+    attention_pattern: torch.Tensor,  # [q_pos k_pos]
+    detection_pattern: torch.Tensor,  # [seq_len seq_len] (seq_len == q_pos == k_pos)
+    *,
+    exclude_bos: bool,
+    exclude_current_token: bool,
+    error_measure: ErrorMeasure,
+) -> float:
+    """Compute the similarity between `attention_pattern` and `detection_pattern`.
+    
+    Args:
+      attention_pattern: Lower triangular matrix (Tensor) representing the attention pattern of a particular attention head.
+      detection_pattern: Lower triangular matrix (Tensor) representing the attention pattern we are looking for.
+      exclude_bos: `True` if the beginning-of-sentence (BOS) token should be omitted from comparison. `False` otherwise. 
+      exclude_bcurrent_token: `True` if the current token at each position should be omitted from comparison. `False` otherwise. 
+      error_measure: "abs" for using absolute values of element-wise differences as the error measure. "mul" for using element-wise multiplication (legacy code).
+    """
+    assert is_square(attention_pattern), f"Attention pattern is not square; got shape {attention_pattern.shape}"
+
+    # mul
+    
+    if error_measure == "mul":
+        if exclude_bos:
```

Comment:
Since exclude_bos and exclude_current_token are the same here and in line 232, these can just occur once at the start of the function.

Reply:
They operate on different tensors, depending on whether `error_measure` is `"mul"` or `"abs"`. I don't see a clear way of merging them.
The best I can come up with is to delegate them to some external function `_exclude_bos_or_current_token` or something but it is so specific that it would probably make the code more obscure instead of clarifying it.